sentry/src/routes/campaign.rs

```diff
@@ -20,25 +38,110 @@ pub async fn create_campaign<A: Adapter>(
-
-    //         match error {
-    //             PoolError::Backend(error) if error.code() == Some(&SqlState::UNIQUE_VIOLATION) => {
-    //                 Err(ResponseError::Conflict(
-    //                     "channel already exists".to_string(),
-    //                 ))
-    //             }
-    //             _ => Err(error_response),
-    //         }
-    //     }
-    //     Ok(false) => Err(error_response),
-    //     _ => Ok(()),
-    // }?;
+
+    match insert_or_modify_campaign(&app.pool, &campaign, &app.redis).await {
+        Err(error) => {
+            // error!(&app.logger, "{}", &error; "module" => "create_channel");
+            return Err(ResponseError::Conflict("channel already exists".to_string()));
+        }
+        Ok(false) => Err(error_response),
+        _ => Ok(()),
+    }?;
 
     let create_response = SuccessResponse { success: true };
 
     Ok(success_response(serde_json::to_string(&campaign)?))
 }
+
+// TODO: Double check redis calls
+async fn get_spent_for_campaign(redis: &MultiplexedConnection, id: CampaignId) -> Result<UnifiedNum, PoolError> {
+    let key = format!("adexCampaign:campaignSpent:{}", id);
+    // campaignSpent tracks the portion of the budget which has already been spent
+    let campaign_spent = match redis::cmd("GET")
+    .arg(&key)
+    .query_async::<_, Option<String>>(&mut redis.clone())
+    .await?
+    {
+        Some(spent) => UnifiedNum::from(spent),
+        // TODO: Double check if this is true
+        // If the campaign is just being inserted, there would be no entry therefore no funds would be spent
+        None => UnifiedNum::from(0)
+    };
+    Ok(campaign_spent)
+}
+
+async fn update_remaining_for_campaign(redis: &MultiplexedConnection, id: CampaignId, amount: UnifiedNum) -> Result<bool, PoolError> {
+    // update a key in Redis for the remaining spendable amount
+    let key = format!("adexCampaign:remainingSpendable:{}", id);
+    redis::cmd("SET")
+        .arg(&key)
+        .arg(amount.to_u64())
+        .query_async(&mut redis.clone())
+        .await?;
+    Ok(true)
+}
+
+async fn update_remaining_for_channel(redis: &MultiplexedConnection, id: ChannelId, amount: UnifiedNum) -> Result<bool, PoolError> {
+    let key = format!("adexChannel:remaining:{}", id);
+    redis::cmd("SET")
+        .arg(&key)
+        .arg(amount.to_u64())
+        .query_async(&mut redis.clone())
+        .await?;
+    Ok(true)
+}
+
+async fn get_campaigns_remaining_sum(redis: &MultiplexedConnection, pool: &DbPool, campaign: &Campaign) -> Result<UnifiedNum, PoolError> {
+    let campaigns_for_channel = get_campaigns_for_channel(&pool, &campaign).await?;
+    let sum_of_campaigns_remaining = campaigns_for_channel
+        .into_iter()
+        .map(async |c| {
+            let spent = get_spent_for_campaign(&redis, c.id).await?;
+            let remaining = c.budget - spent;
+            remaining
+        })
+        .sum();
```

Comment:
This `sum()` should return `Option<UnifiedNum>` as it might overflow.
We need to make it a `Result` and handle it